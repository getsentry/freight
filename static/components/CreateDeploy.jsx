import {browserHistory} from 'react-router';
import jQuery from 'jquery';
import PropTypes from 'prop-types';
import React from 'react';

import api from '../api';

class CreateDeploy extends React.Component {
  static propTypes = {
    appList: PropTypes.array,
  };

  constructor(props) {
    super(props);
    const appList = props.appList;
    const defaultApp = appList.length !== 0 ? appList[0] : null;
    const app = props.location.query?.app
      ? props.location.query.app
      : defaultApp
      ? defaultApp.name
      : null;
    const envMap = app
      ? appList.filter(appObj => appObj.name === app)[0].environments || {}
      : {};
    const envList = Object.keys(envMap);
    const env = envList.length ? envList[0] : null;
    const defaultRef = env ? envMap[env].defaultRef : 'master';

    this.state = {
      app,
      env,
      envMap,
      ref: defaultRef,
      submitInProgress: false,
      submitError: null,
    };
  }

  onChangeApplication = e => {
    const val = jQuery(e.target).val();
    const envMap = val
      ? this.props.appList.filter(app => {
          return app.name === val;
        })[0].environments || {}
      : {};
    const envList = Object.keys(envMap);
    const env = envList.length ? envList[0] : null;
    this.setState({
      app: val,
      envMap,
      env,
      ref: env ? env.defaultRef : 'master',
    });
  };

  onChangeEnvironment = e => {
    const val = jQuery(e.target).val();
    const config = val ? this.state.envMap[val] || {} : {};
    this.setState({
      env: val,
      ref: config.defaultRef || 'master',
    });
  };

  onChangeRef = e => {
    this.setState({
      ref: e.target.value,
    });
  };

  onSubmit = e => {
    e.preventDefault();

    if (this.state.submitInProgress) {
      return;
    }

    this.setState(
      {
        submitInProgress: true,
      },
      () => {
        api.request('/deploys/', {
          method: 'POST',
          data: {
            app: this.state.app,
            env: this.state.env,
            ref: this.state.ref,
          },
          success: data => {
            this.gotoDeploy(data);
          },
          error: response => {
            this.setState({
              submitError: response.responseJSON.error,
              submitInProgress: false,
            });
          },
        });
      }
    );
  };

  gotoDeploy = deploy => {
    const {app, environment, number} = deploy;
    browserHistory.push(`/deploys/${app.name}/${environment}/${number}`);
  };

  render() {
    return (
      <div className="create-deploy">
        <div className="section">
          <div className="section-header">
            <h2>Create Deploy</h2>
          </div>
          {this.state.submitError && (
            <div className="alert alert-block alert-danger">
              <strong>Unable to create deploy:</strong>
              <span>{this.state.submitError}</span>
            </div>
          )}
          <form onSubmit={this.onSubmit}>
            <div className="form-group">
              <label>Application</label>
              <select
                className="form-control"
                value={this.state.app}
                onChange={this.onChangeApplication}
              >
                {this.props.appList.map(app => {
                  return <option key={app.name}>{app.name}</option>;
                })}
              </select>
            </div>
            <div className="form-group">
              <label>Environment</label>
              <select
                className="form-control"
                value={this.state.env}
                onChange={this.onChangeEnvironment}
              >
                {Object.keys(this.state.envMap).map(env => {
                  return <option key={env}>{env}</option>;
                })}
              </select>
            </div>
            <div className="form-group">
              <label>Ref</label>
              <input
                type="text"
                className="form-control"
                onChange={this.onChangeRef}
                placeholder="e.g. master"
                value={this.state.ref}
              />
            </div>
            <div className="submit-group">
              <button
                type="submit"
                className="btn btn-primary"
                disabled={this.state.submitInProgress}
              >
                Ship It!
              </button>
            </div>
          </form>
        </div>
      </div>
    );
  }
}

export default CreateDeploy;
