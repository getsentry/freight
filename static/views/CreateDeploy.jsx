import * as React from 'react';
import {browserHistory} from 'react-router';

import api from 'app/api';

function gotoDeploy(deploy) {
  const {app, environment, number} = deploy;
  browserHistory.push(`/deploys/${app.name}/${environment}/${number}`);
}

function CreateDeploy({location, appList = []}) {
  const firstApp = appList.length !== 0 ? appList[0] : null;

  const defaultApp = location.query?.app
    ? location.query.app
    : firstApp
    ? firstApp.name
    : null;

  const [app, setApp] = React.useState(defaultApp);

  const envMap = React.useMemo(
    () => appList.find(appObj => appObj.name === app)?.environments ?? {},
    [app, appList]
  );

  const envList = React.useMemo(() => Object.keys(envMap), [envMap]);

  const [env, setEnv] = React.useState(envList?.[0] ?? null);
  const [ref, setRef] = React.useState(envMap[env]?.defaultRef ?? 'master');

  // update environment when environment list changes
  React.useEffect(() => void setEnv(envList?.[0] ?? null), [envList]);

  // Update refs when app or env changes
  React.useEffect(
    () => void setRef(envMap?.[env]?.defaultRef ?? 'master'),
    [env, envMap]
  );

  const [submitError, setSubmitError] = React.useState(false);
  const [submitInProgress, setSubmitInProgress] = React.useState(false);

  const startDeploy = React.useCallback(async () => {
    const deployResp = await api.request('/deploys/', {
      method: 'POST',
      data: {app, env, ref},
    });

    const result = await deployResp.json();

    if (deployResp.ok) {
      gotoDeploy(result);
    } else {
      setSubmitError(result.error);
      setSubmitInProgress(false);
    }
  }, [app, env, ref]);

  /**
   * Handles creating a deploy
   */
  const handleSubmit = React.useCallback(
    e => {
      e.preventDefault();

      if (submitInProgress) {
        return;
      }

      setSubmitInProgress(true);
      startDeploy();
    },
    [startDeploy, submitInProgress]
  );

  return (
    <div className="create-deploy">
      <div className="section">
        <div className="section-header">
          <h2>Create Deploy</h2>
        </div>
        {submitError && (
          <div className="alert alert-block alert-danger">
            <strong>Unable to create deploy: </strong>
            <span>{submitError}</span>
          </div>
        )}
        <form onSubmit={handleSubmit}>
          <div className="form-group">
            <label>Application</label>
            <select
              className="form-control"
              value={app}
              onChange={e => setApp(e.target.value)}
            >
              {appList.map(a => (
                <option key={a.name}>{a.name}</option>
              ))}
            </select>
          </div>
          <div className="form-group">
            <label>Environment</label>
            <select
              className="form-control"
              value={env}
              onChange={e => setEnv(e.target.value)}
            >
              {Object.keys(envMap).map(e => (
                <option key={e}>{e}</option>
              ))}
            </select>
          </div>
          <div className="form-group">
            <label>Ref</label>
            <input
              type="text"
              className="form-control"
              onChange={e => setRef(e.target.value)}
              placeholder="e.g. master"
              value={ref}
            />
          </div>
          <div className="submit-group">
            <button type="submit" className="btn btn-primary" disabled={submitInProgress}>
              Ship It!
            </button>
          </div>
        </form>
      </div>
    </div>
  );
}

export default CreateDeploy;
