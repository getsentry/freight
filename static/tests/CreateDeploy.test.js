import React from 'react';
import {mount} from 'enzyme';

import CreateDeploy from '../components/CreateDeploy.jsx';

test('CreateDeploy Snapshot', () => {
  const appList = [
    {
      environments: {
        production: {
          defaultRef: 'master',
        },
        staging: {
          defaultRef: 'HEAD',
        },
      },
      id: '1',
      name: 'freight',
      repository: 'https://github.com/getsentry/freight.git',
    },
  ];
  const wrapper = mount(<CreateDeploy appList={appList} location={{}} />);
  expect(wrapper).toMatchSnapshot();
});
