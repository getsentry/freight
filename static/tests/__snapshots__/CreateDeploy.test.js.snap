// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateDeploy Snapshot 1`] = `
<CreateDeploy
  appList={
    Array [
      Object {
        "environments": Object {
          "production": Object {
            "defaultRef": "master",
          },
          "staging": Object {
            "defaultRef": "HEAD",
          },
        },
        "id": "1",
        "name": "freight",
        "repository": "https://github.com/getsentry/freight.git",
      },
    ]
  }
  location={Object {}}
>
  <div
    className="create-deploy"
  >
    <div
      className="section"
    >
      <div
        className="section-header"
      >
        <h2>
          Create Deploy
        </h2>
      </div>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label>
            Application
          </label>
          <select
            className="form-control"
            onChange={[Function]}
            value="freight"
          >
            <option
              key="freight"
            >
              freight
            </option>
          </select>
        </div>
        <div
          className="form-group"
        >
          <label>
            Environment
          </label>
          <select
            className="form-control"
            onChange={[Function]}
            value="production"
          >
            <option
              key="production"
            >
              production
            </option>
            <option
              key="staging"
            >
              staging
            </option>
          </select>
        </div>
        <div
          className="form-group"
        >
          <label>
            Ref
          </label>
          <input
            className="form-control"
            onChange={[Function]}
            placeholder="e.g. master"
            type="text"
            value="master"
          />
        </div>
        <div
          className="submit-group"
        >
          <button
            className="btn btn-primary"
            disabled={false}
            type="submit"
          >
            Ship It!
          </button>
        </div>
      </form>
    </div>
  </div>
</CreateDeploy>
`;
